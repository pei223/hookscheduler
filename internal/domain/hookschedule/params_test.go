package hookschedule_test

import (
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/google/uuid"
	"github.com/pei223/hook-scheduler/internal/domain/hookschedule"
	"github.com/pei223/hook-scheduler/internal/test_common"
	"github.com/samber/lo"
	"github.com/stretchr/testify/require"
)

func TestValidation(t *testing.T) {
	baseParams := hookschedule.HookScheduleCreateParams{
		HookID:                uuid.New(),
		DisplayName:           "test",
		Description:           "test schedule",
		ScheduleFrequencyUnit: hookschedule.ScheduleFrequencyEveryMinute,
		ScheduleTimeMonth:     1,
		ScheduleTimeDay:       1,
		ScheduleTimeHour:      0,
		ScheduleTimeMinute:    0,
		ScheduleTimeSecond:    0,
	}

	t.Run("success", func(t *testing.T) {
		invalidParams := baseParams.Validate()
		require.Nil(t, invalidParams)
	})

	badRequestCases := []struct {
		caseName     string
		invalidCount int
		genBody      func(base hookschedule.HookScheduleCreateParams) hookschedule.HookScheduleCreateParams
	}{
		{
			caseName:     "too long displayName",
			invalidCount: 1,
			genBody: func(base hookschedule.HookScheduleCreateParams) hookschedule.HookScheduleCreateParams {
				base.DisplayName = "1234567890123456789012345678901234567890"
				return base
			},
		},
		{
			caseName:     "empty displayName",
			invalidCount: 1,
			genBody: func(base hookschedule.HookScheduleCreateParams) hookschedule.HookScheduleCreateParams {
				base.DisplayName = ""
				return base
			},
		},
		{
			caseName:     "invalid scheduleTimeMonth",
			invalidCount: 1,
			genBody: func(base hookschedule.HookScheduleCreateParams) hookschedule.HookScheduleCreateParams {
				base.ScheduleTimeMonth = 13
				return base
			},
		},
		{
			caseName:     "invalid scheduleTimeDay",
			invalidCount: 1,
			genBody: func(base hookschedule.HookScheduleCreateParams) hookschedule.HookScheduleCreateParams {
				base.ScheduleTimeDay = 32
				return base
			},
		},
		{
			caseName:     "invalid scheduleTimeHour",
			invalidCount: 1,
			genBody: func(base hookschedule.HookScheduleCreateParams) hookschedule.HookScheduleCreateParams {
				base.ScheduleTimeHour = 24
				return base
			},
		},
		{
			caseName:     "invalid scheduleTimeMinute",
			invalidCount: 1,
			genBody: func(base hookschedule.HookScheduleCreateParams) hookschedule.HookScheduleCreateParams {
				base.ScheduleTimeMinute = 60
				return base
			},
		},
		{
			caseName:     "invalid scheduleTimeSecond",
			invalidCount: 1,
			genBody: func(base hookschedule.HookScheduleCreateParams) hookschedule.HookScheduleCreateParams {
				base.ScheduleTimeSecond = 60
				return base
			},
		},
		{
			caseName:     "invalid scheduleFrequencyUnit",
			invalidCount: 1,
			genBody: func(base hookschedule.HookScheduleCreateParams) hookschedule.HookScheduleCreateParams {
				base.ScheduleFrequencyUnit = "invalid"
				return base
			},
		},
	}

	for _, c := range badRequestCases {
		t.Run(c.caseName, func(t *testing.T) {
			baseBody := lo.Must(test_common.DeepCopy(baseParams))
			body := c.genBody(*baseBody)

			invalidParams := body.Validate()
			require.Len(t, invalidParams, c.invalidCount)
			snaps.MatchJSON(t, invalidParams)
		})
	}
}
