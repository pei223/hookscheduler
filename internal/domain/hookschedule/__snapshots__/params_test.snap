
[TestValidation/too_long_displayName - 1]
[
 {
  "name": "displayName",
  "reason": "DisplayName must be a maximum of 20 characters in length"
 }
]
---

[TestValidation/empty_displayName - 1]
[
 {
  "name": "displayName",
  "reason": "DisplayName is a required field"
 }
]
---

[TestValidation/invalid_scheduleTimeMonth - 1]
[
 {
  "name": "scheduleTimeMonth",
  "reason": "ScheduleTimeMonth must be 12 or less"
 }
]
---

[TestValidation/invalid_scheduleTimeDay - 1]
[
 {
  "name": "scheduleTimeDay",
  "reason": "ScheduleTimeDay must be 31 or less"
 }
]
---

[TestValidation/invalid_scheduleTimeHour - 1]
[
 {
  "name": "scheduleTimeHour",
  "reason": "ScheduleTimeHour must be 23 or less"
 }
]
---

[TestValidation/invalid_scheduleTimeMinute - 1]
[
 {
  "name": "scheduleTimeMinute",
  "reason": "ScheduleTimeMinute must be 59 or less"
 }
]
---

[TestValidation/invalid_scheduleTimeSecond - 1]
[
 {
  "name": "scheduleTimeSecond",
  "reason": "ScheduleTimeSecond must be 59 or less"
 }
]
---

[TestValidation/invalid_scheduleIntervalUnit - 1]
[
 {
  "name": "scheduleFrequencyUnit",
  "reason": "ScheduleFrequencyUnit must be one of [1 2 3 5 6]"
 }
]
---

[TestValidation/invalid_scheduleFrequencyUnit - 1]
[
 {
  "name": "scheduleFrequencyUnit",
  "reason": "ScheduleFrequencyUnit must be one of [every_minute every_hour every_day every_month every_year]"
 }
]
---
